cudax/include/cuda/experimental/__execution/bulk.cuh

```diff
@@ -0,0 +1,479 @@
+    // operation state type.
+    template <class _CvSndr, class _Rcvr>
+    using __opstate_t = typename _BulkTag::template __opstate_t<_CvSndr, _Rcvr, _Shape, _Fn>;
+
+    template <class _Self, class... _Env>
+    [[nodiscard]] _CCCL_API static _CCCL_CONSTEVAL auto get_completion_signatures()
+    {
+      _CUDAX_LET_COMPLETIONS(
+        auto(__child_completions) = execution::get_child_completion_signatures<_Self, _Sndr, _Env...>())
+      {
+        return transform_completion_signatures(__child_completions, __transform_value_completion_fn<_Shape, _Fn>{});
+      }
+    }
+
+    // The bulk algorithm lowers to a bulk_chunked sender. The bulk sender itself should
+    // not have `connect` functions, shince they should never be called. Hence, we
+    // constrain these functions with !same_as<_BulkTag, bulk_t>.
+    _CCCL_TEMPLATE(class _Rcvr)
+    _CCCL_REQUIRES((!_CUDA_VSTD::same_as<_BulkTag, bulk_t>) )
+    [[nodiscard]] _CCCL_API constexpr auto connect(_Rcvr __rcvr) && -> __opstate_t<_Sndr, _Rcvr>
+    {
+      return __opstate_t<_Sndr, _Rcvr>{
+        static_cast<_Sndr&&>(__sndr_),
+        static_cast<_Rcvr&&>(__rcvr),
+        __state_.__shape_,
+        static_cast<_Fn&&>(__state_.__fn_)};
+    }
+
+    _CCCL_TEMPLATE(class _Rcvr)
+    _CCCL_REQUIRES((!_CUDA_VSTD::same_as<_BulkTag, bulk_t>) )
+    [[nodiscard]] _CCCL_API constexpr auto connect(_Rcvr __rcvr) const& -> __opstate_t<const _Sndr&, _Rcvr>
+    {
+      return __opstate_t<const _Sndr&, _Rcvr>{__sndr_, static_cast<_Rcvr&&>(__rcvr), __state_.__shape_, __state_.__fn_};
+    }
+
+    [[nodiscard]] _CCCL_API constexpr auto get_env() const noexcept -> __bulk_attrs_t<_Sndr, _Shape>
+    {
+      return {__state_.__shape_, __sndr_};
+    }
+
+    _CCCL_NO_UNIQUE_ADDRESS _BulkTag __tag_;
+    __bulk_state_t<_Policy, _Shape, _Fn> __state_;
+    _Sndr __sndr_;
+  };
+
+  template <class _Policy, class _Shape, class _Fn>
+  struct _CCCL_TYPE_VISIBILITY_DEFAULT __closure_t : __bulk_state_t<_Policy, _Shape, _Fn>
+  {
+    template <class _Sndr>
+    _CCCL_TRIVIAL_API friend constexpr auto operator|(_Sndr&& __sndr, __closure_t __self)
+    {
+      using __domain_t = __early_domain_of_t<_Sndr>;
+      using __sndr_t   = __bulk_t::__sndr_t<_Sndr, _Policy, _Shape, _Fn>;
+      return transform_sender(__domain_t{},
+                              __sndr_t{{}, static_cast<__closure_t&&>(__self), static_cast<_Sndr&&>(__sndr)});
+    }
+  };
+
+  // This function call operator is the entry point for the bulk algorithms. It takes a
+  // predecessor sender, a policy, a shape, and a function, and returns a sender that can
+  // be connected to a receiver.
+  template <class _Sndr, class _Policy, class _Shape, class _Fn>
+  _CCCL_API auto operator()(_Sndr&& __sndr, _Policy __policy, _Shape __shape, _Fn __fn) const
+  {
+    static_assert(__is_sender<_Sndr>);
+    static_assert(_CUDA_VSTD::integral<_Shape>);
+    static_assert(is_execution_policy_v<_Policy>);
```

Comment:
To be sure, those are "Mandates" not "Constraints"

Reply:
yes